packages/oidc/src/token.ts

```diff
@@ -0,0 +1,226 @@
+    if (!fs.existsSync(pkgPath)) {
+      throw new Error('project.json not found');
+    }
+    const pkg = JSON.parse(fs.readFileSync(pkgPath, 'utf8'));
+    return pkg.projectId;
+  } catch (e) {
+    const message = e instanceof Error ? e.message : e;
+    throw new Error(`unable to find project id: ${message}`);
+  }
+}
+
+async function findRootDir(): Promise<string> {
+  try {
+    const fs = await import('fs');
+    const path = await import('path');
+    let dir = process.cwd();
+    while (dir !== path.dirname(dir)) {
+      const pkgPath = path.join(dir, '.vercel');
+      if (fs.existsSync(pkgPath)) {
+        return dir;
+      }
+      dir = path.dirname(dir);
+    }
+  } catch (e) {
+    throw new Error('token refresh only supported in node server environments');
+  }
+  throw new Error('unable to find root directory');
+}
+
+async function saveToken(
+  token: VercelTokenResponse,
+  projectId: string
+): Promise<void> {
+  try {
+    const os = await import('os');
+    const path = await import('path');
+    const tmpDir = os.tmpdir();
+    const tokenPath = path.join(
+      tmpDir,
+      `vercel-oidc-token-${projectId}`,
+      'token.json'
+    );
+    const tokenJson = JSON.stringify(token);
+    const fs = await import('fs');
+    fs.mkdirSync(path.dirname(tokenPath), { recursive: true });
+    fs.writeFileSync(tokenPath, tokenJson);
+    return;
+  } catch (e) {
+    const message = e instanceof Error ? e.message : e;
+    throw new Error(`failed to save token: ${message}`);
+  }
+}
+
+async function loadToken(
+  projectId: string
+): Promise<VercelTokenResponse | null> {
+  try {
+    const os = await import('os');
+    const path = await import('path');
+    const fs = await import('fs');
+
+    const tmpDir = os.tmpdir();
+    const tokenPath = path.join(
+      tmpDir,
+      `vercel-oidc-token-${projectId}`,
+      'token.json'
+    );
+    if (!fs.existsSync(tokenPath)) {
+      return null;
+    }
+    const tokenJson = fs.readFileSync(tokenPath, 'utf8');
+    const token = JSON.parse(tokenJson) as VercelTokenResponse;
+    return token;
+  } catch (e) {
+    const message = e instanceof Error ? e.message : e;
+    throw new Error(`failed to load token: ${message}`);
+  }
+}
+
+export function getTokenPayload(token: string): TokenPayload {
+  const tokenParts = token.split('.');
+  if (tokenParts.length !== 3) {
+    throw new Error('Invalid token');
+  }
+
+  const base64 = tokenParts[1].replace(/-/g, '+').replace(/_/g, '/');
+
+  if (isBrowserEnv()) {
+    if (typeof Uint8Array === undefined || typeof TextDecoder === undefined) {
+      throw new Error(
+        'The browser environment is not supported. Requires Uint8Array and TextDecoder'
+      );
+    }
+
+    const binaryString = window.atob(base64);
+    const toUtf8 = Uint8Array.from(binaryString, c => c.charCodeAt(0));
+    const decoder = new TextDecoder();
+    const payload = decoder.decode(toUtf8);
```

Comment:
I think there may be some alternative Node.js APIs you can use for doing a Base64URL-decoding. E.g.,

```js
function decodeBase64Url(base64url) {
  const base64 = base64url
    .replace(/-/g, '+')
    .replace(/_/g, '/') + '=='.slice(0, (4 - (base64url.length % 4)) % 4))

  return Buffer.from(base64, 'base64')
}
```

Is that an option? Or are we trying to run in other runtimes where these are unavailable?